BaseLib/quicksort.h

```diff
@@ -15,243 +12,97 @@
+void quicksort(std::vector<T1>& array, std::size_t beg, std::size_t end, std::vector<T2>& perm)
 {
-	if (beg < end) {
-		std::size_t p = partition_(array, beg, end, second_array);
-		quicksort(array, beg, p, second_array);
-		quicksort(array, p + 1, end, second_array);
+	// Zip input arrays.
+	std::vector<std::pair<T1, T2>> data;
+	data.reserve(array.size());
+	std::transform(array.begin(), array.end(), perm.begin(),
+		std::back_inserter(data),
+		[](T1 const& t1, T2 const& t2)
+		{
+			return std::make_pair(t1, t2);
+		});
+
+	// Sort data using first element of the pair.
+	std::sort(data.begin(), data.end(),
+		[](std::pair<T1, T2> const& a, std::pair<T1, T2> const& b)
+		{
+			return (a.first < b.first);
+		});
+
+	// Unzip sorted data.
+	for (std::size_t i = 0; i < data.size(); i++)
+	{
+		array[i] = data[i].first;
+		perm[i] = data[i].second;
 	}
 }
 
-} // end namespace BaseLib
-
-// STL
-#include <vector>
-
-namespace BaseLib
-{
 template <typename T1, typename T2 = std::size_t>
-class Quicksort
+void quicksort(std::vector<T1*>& array, std::size_t beg, std::size_t end, std::vector<T2>& perm)
 {
-public:
-	Quicksort (std::vector<T1>& array, std::size_t beg, std::size_t end, std::vector<T2>& perm)
-	{
-		quicksort (array, beg, end, perm);
-	}
-private:
-	std::size_t partition_(std::vector<T1>& array,
-	                       std::size_t beg,
-	                       std::size_t end,
-	                       std::vector<T2>& perm)
-	{
-		std::size_t i = beg + 1;
-		std::size_t j = end - 1;
-		T1 m = array[beg];
-
-		for (;;) {
-			while ((i < end) && (array[i] <= m))
-				i++;
-			while ((j > beg) && !(array[j] <= m))
-				j--;
-
-			if (i >= j)
-				break;
-			std::swap(array[i], array[j]);
-			std::swap(perm[i], perm[j]);
-		}
-
-		std::swap(array[beg], array[j]);
-		std::swap(perm[beg], perm[j]);
-		return j;
-	}
-
-	void quicksort(std::vector<T1>& array,
-	               std::size_t beg,
-	               std::size_t end,
-	               std::vector<T2>& perm)
+	// Zip input arrays.
+	std::vector<std::pair<T1*, T2>> data;
+	data.reserve(array.size());
+	std::transform(array.begin(), array.end(), perm.begin(),
+		std::back_inserter(data),
+		[](T1* const& t1, T2 const& t2)
+		{
+			return std::make_pair(t1, t2);
+		});
+
+	// Sort data using first element of the pair.
+	std::sort(data.begin(), data.end(),
+		[](std::pair<T1*, T2> const& a, std::pair<T1*, T2> const& b)
+		{
+			return (*a.first < *b.first);
+		});
+
+	// Unzip sorted data.
+	for (std::size_t i = 0; i < data.size(); i++)
 	{
-		if (beg < end) {
-			std::size_t p = partition_(array, beg, end, perm);
-			quicksort(array, beg, p, perm);
-			quicksort(array, p + 1, end, perm);
-		}
+		array[i] = data[i].first;
+		perm[i] = data[i].second;
 	}
-};
+}
 
-// specialization for pointer types
-template <typename T1, typename T2>
-class Quicksort <T1*, T2>
+template <typename T1, typename T2 = std::size_t>
+void quicksort(T1* & array, std::size_t beg, std::size_t end, T2* & perm)
```

Comment:
i cannot find it. as far as I checked, previously this was used  in AdjMat.cpp but you have changed it to use std::sort().